File: 2021/day3/part1.py
with open("day3\\input.txt", 'r') as f:
    lines = f.readlines()

gamma_rate = ""
epsilon_rate = 0

for i in range(len(lines[0])-1):
    count_0 = 0
    count_1 = 0

    for line in lines:
        if line[i] == "0":
            count_0 += 1
        elif line[i] == "1":
            count_1 += 1

    if count_0 > count_1:
        print("higher rate: 0")
        gamma_rate += "0"
    else:
        print("higher rate: 1")
        gamma_rate += "1"

gamma_rate = int(gamma_rate, 2)
epsilon_rate = gamma_rate ^ (2**(len(lines[0])-1) - 1)

print(gamma_rate, epsilon_rate)
print(gamma_rate * epsilon_rate)

File: 2021/day3/part1.r
dataset <- do.call(rbind, lapply(
    strsplit(scan("day3\\input.txt", what = ""), ""), as.numeric
))

data_size <- length(dataset[1, ])

gamma_rate <- 0

for (i in 1:data_size) {
    col <- dataset[, i]

    count_0 <- length(col[col == 0])
    count_1 <- length(col[col == 1])

    if (count_0 > count_1) {
        cat("higher rate: 0\n")
        gamma_rate <- gamma_rate * 2
    } else {
        cat("higher rate: 1\n")
        gamma_rate <- gamma_rate * 2 + 1
    }
}

epsilon_rate <- bitwXor(gamma_rate, ((2**data_size) - 1))

cat(paste(
    "gamma rate: ", gamma_rate, "\n",
    "epsilon rate: ", epsilon_rate, "\n",
    sep = ""))
cat(paste("final answer: ", gamma_rate * epsilon_rate, sep = ""))
